### ISSUE TITLE
Optimizers in the C++ API

### ISSUE THREAD

**Issue by theflofly** [tensorflow/tensorflow#9837]
There is currently no Optimizer in the [C++ API](https://www.tensorflow.org/api_docs/cc/) compared as the ones that we can find in the [Python API](https://www.tensorflow.org/api_docs/python/).

It means, when using only the C++ API that we have to manually collect the gradients and apply them.

Don't you think it should be a convenient add? Retrieving the gradients and applying them is kind of hard and error prone. If yes, I could work on it and create an Optimizer + GradientDescentOptimizer in the C++ API.

I maybe missed something.

**Comment by drpngx** [comment]
I would recommend working with us to define the interface and design that looks like [gradients](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/cc/gradients). That would be a nice contribution!

Ideally, we would have  a check that things don't go out of sync with python.

**Comment by asimshankar** [comment]
Yes, the C++ (and other language) APIs lack many of the higher level Python conveniences right now.
(Side note: The gradient registry for all operations in C++ is not as comprehensive as Python, so there is work to be done there - for details see the link that @drpngx pointed to).

The TensorFlow team doesn't have the cycles to pursue this ourselves right now, but if you'd like to give it a shot we'd be happy to discuss any interfaces/design that you propose.

FYI @suharshs @skye @josh11b

**Comment by theflofly** [comment]
Ok. I'll see how I can do things and I'll make proposals here. I'll surely contribute to the C++ gradients as you pointed it out.

**Comment by theflofly** [comment]
My observations until now.

In python the Optimizer class provides a minimize method that calls compute_gradient and apply_gradient. compute_gradient adds the gradient ops to the graph, apply_gradient adds the apply ops to the graph and return an operation that applies the gradients.

The mains files are:
* optimizer.py
* gradient_descent.py
* gradients_impl.py

In C++ some gradient operations are implemented. There is also a SymbolicGradientBuilder that applies the gradient operations to the graph. The missing part is the applying of the gradients.

The steps would be:
* Create the Optimizer and GradientDescentOptimizer class
* Use the SymbolicGradientBuilder in the ComputeGradient method
* Add apply operations to the graph and return an operation in the ApplyGradient method
* Combine both methods in the minimize method
* Unit tests

Any insight?

**Comment by AhmetCanSolak** [comment]
Hi @theflofly ! Your steps sounds reasonable in general. I would like to work on that but need guidance to start. Please contact if you are still interested in.

**Comment by theflofly** [comment]
Hi @AhmetCanSolak, I am working on it and I made good progress even if it is not as easy as it sounds. 

I don't see a way to split the task unfortunately. Especially at this point. Once the optimizer part is added, there will be plenty of work. Namely adding all the Optimizers (adam, adagrad, momentum, etc.). If you want I can ping you at this time and we'll share the work.

In the meantime, there is always the gradient operations that need to be ported.

**Comment by AhmetCanSolak** [comment]
Great. It is better idea to ping me when sharing tasks possible. Meanwhile I better check again the contribution guidelines since I have not contributed to TF yet.

**Comment by Moriadry** [comment]
@theflofly Do you have any ideas about bringing automatic differentiation (gradients) to C++ API? 
also, If any help needed, you can ping me:)

**Comment by theflofly** [comment]
@Moriadry AD is already here.

```
std::vector<Output> grad_outputs;
TF_ASSERT_OK( AddSymbolicGradients(scope, {z}, {x, y}, {dz}, &grad_outputs));
```
Will add the gradient operations to the graph. You can find more example here. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/cc/framework/gradients_test.cc

The goal of this issue is mainly about the update part. If you want to add some gradients you can follow these [guidelines](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/cc/gradients).

Does this answer?

**Comment by Moriadry** [comment]
@theflofly ok, your answer does make sense, thank you!

**Comment by theflofly** [comment]
Hello, this is a message to keep you updated. I think that I have a working version, I'll soon share the code. I still have a TODO list of minor things to do.

**Comment by Moriadry** [comment]
@theflofly happy to hear that, looking forward to your code.

**Comment by theflofly** [comment]
I opened a pull request. Feel free to share your thoughts.

It basically add this kind of operation:
```auto train = GradientDescentOptimizer(0.01).Minimize(scope, {loss});```

You can see a detailed example here: https://github.com/theflofly/tensorflow/blob/optimizer-cc-issue-9837/tensorflow/cc/training/optimizer_test.cc in the second function.

I put some TODO(theflofly) listing some improvements. I will add others optimizers and some gradients operations, then I'll work on these TODOs. 

We can easily share the work, so if someone wants to work either on the TODOs, gradients or optimizers please create an issue and ping me so that we don't work on the same feature.

**Comment by fferroni** [comment]
Presumably to achieve this you need to have created a function similar to tf.gradients() right? Is there any plan on exposing this too in C++ ?

**Comment by theflofly** [comment]
@fferroni Indeed. The function is already existing and was created by someone at Google. In this issue I basically wrap existing stuffs into a cleaner API similar to the python.

I already explained above but basically you can use:

```
auto x = Const(scope_, {{1.0f, 2.0f}, {3.0f, 4.0f}, {5.0f, 6.0f}});
auto y = Const(scope_, {{1.0f}, {2.0f}, {3.0f}});

auto w1 = Variable(scope_, {2, 1}, DT_FLOAT);
auto assign_w1 = Assign(scope_, w1, Const(scope_, {{0.1f}, {0.2f}}));

auto layer_1 = Tanh(scope_, MatMul(scope_, x, w1));

std::vector<Output> grad_outputs;
TF_CHECK_OK(AddSymbolicGradients(scope_, {layer_1}, {assign_w1}, &grad_outputs));

auto apply = ApplyGradientDescent(
      scope_, {w1},
      Cast(scope_, 0,01,  static_cast<DataType>(w1.type() - 100)),
      {grad_outputs[0]});

ClientSession session(scope_);
TF_CHECK_OK(session.Run({assign_w1}, nullptr));
TF_CHECK_OK(session.Run({apply}, nullptr));

```
The equivalent to gradients from python is AddSymbolicGradients. Not sure the above code is working as I can't test it, but you get the idea on how to put things together.

Keep in mind that you first have to call session run on assign nodes, also the AddSymbolicGradients will not work if you pass the var w1 and not the assign node whereas to apply gradient descent you have to pass the var node and not the assign node. I edited AddSymbolicGradients on my PR to solve this issue.

With more than one var the above code increase and complexity. That's why I replaced it by: `auto train = GradientDescentOptimizer(0.01).Minimize(scope, {loss});`.

**Comment by josh11b** [comment]
@asimshankar @skye

**Comment by theflofly** [comment]
FYI this will certainly not be merged, you can see the answer of @asimshankar in https://github.com/tensorflow/tensorflow/pull/11377.

**Comment by dguerra** [comment]
So how can it be possible to add gradients manually now to a graph?

I can run and train a linear regression graph when I compute the gradients myself.

```
#include "tensorflow/cc/client/client_session.h"
#include "tensorflow/cc/ops/standard_ops.h"
#include "tensorflow/cc/framework/gradients.h"
#include "tensorflow/core/framework/tensor.h"

int main()
{
  using namespace tensorflow;
  using namespace tensorflow::ops;
  Scope root = Scope::NewRootScope();
  
  auto W = Variable(root.WithOpName("W"), {3,1}, DT_DOUBLE);

  auto x = Placeholder(root.WithOpName("x"), DT_DOUBLE);
  auto y = Placeholder(root.WithOpName("y"), DT_DOUBLE);

  auto m = MatMul(root, x, W);
  auto dd = Subtract(root, y, m);
  
 //compute loss
  auto loss = MatMul(root, MatMul(root, dd, dd, MatMul::TransposeA(true)), {0.5}); 
  double learning_rate = 0.1;
 

  std::vector<Output> grad_outputs;
  TF_CHECK_OK(AddSymbolicGradients(root, {loss}, {m}, &grad_outputs));
  //Compute gradients with formula
  //  auto grad_W = Subtract(root, MatMul(root, MatMul(root, x, x, MatMul::TransposeA(true)), W), MatMul(root, x, y,  MatMul::TransposeA(true)));

  auto apply_grad_W = ApplyGradientDescent(root, W, learning_rate,  grad_outputs[0]);

  //Initialize variables
  auto init_W = Assign(root, W, {{1.0},{1.0},{1.0}});

  std::vector<Tensor> outputs;
  ClientSession session(root);

  //Run variable initializers
  session.Run({init_W}, &outputs);

  for(unsigned int i=0;i<200;++i)
  {
    TF_CHECK_OK(session.Run( { {x,{{1.0,-1.0,3.0}, {1.0,2.0,1.0}, {1.0,-2.0,-2.0}, {1.0,0.0,2.0}}}, {y,{{14.0}, {15.0}, {-9.0}, {13.0}}} } , {loss, apply_grad_W}, &outputs));
    std::cout << std::string("loss: ") << outputs[0].scalar<double>() << std::endl << std::string("weights: ")<< outputs[1].matrix<double>() << std::endl;
  }
  return 0;
}

```

But when I try to get the gradients with AddSymbolicGradients and despite MatMulGrad been already defined in math_grad.cc, I got this error message:

"Not found: No gradient defined for op: MatMul. Please see https://www.tensorflow.org/code/tensorflow/cc/gradients/README.md for instructions on how to add C++ gradients."

**Comment by theflofly** [comment]
As the core is in C++ I guess:
* The PR should be merged
* tensorflow/cc/gradients should be moved in tensorflow/core/
* All the python conveniences should be added to tensorflow/core
* The non C++ API should all become bindings to the C++ core code

The problem is that there is a lot of python code to be rewritten (but I would be happy to work on that with the community) but more importantly, the data scientist + researchers are well versed in python, that's why we have so much more features in python.

The C++ is not sexy, less easy and the use of bazel means no proper IDE support for now. Debugging is hard, especially on OS X because of a bazel bug that don't add symbols to the binary generated. I am using VS Code personally, there is a project called tulsi that convert a bazel project to Xcode but it does not work for TF because it does not handle "cc_library" if I remember correctly.

I am waiting for the @asimshankar answer to more insights. But clearly one implementation of the gradient descent and maybe other optimizers in the future for each language is clearly a bad design choice... He talked about a C implementation, but the core is in C++ so I don't understand why.

Let's wait. :)

For your problem, are you using the last version of master ? Because me it gives me: `Not found: No gradient defined for op: Sub.`. But the MatMul is not the problem.

There is currently no gradient for the binary maths operations, so you can't do basic stuff like add, sub, ... I was working on them because I also need them. But I'll wait because duplicating the gradients operations for each languages is also a bad idea.

**Comment by dguerra** [comment]
Definetely I'd like to get more insights as well. I'd like to see a working example of training a basic neural net or convnet with C++ API. If basic gradient ops such as Subtract are not supported yet, I guess there is still a long way to go until that. I see SubGrad is already defined in /tensorflow/core/ops/math_grad.cc, so I don't know why it isn't available from the C++ API yet.

**Comment by theflofly** [comment]
Even if the gradient is defined into the core, it has to be put in the gradient registry in order to be used by the  AddSymbolicGradients method. Basically register grad has to be called, and it is not the case.

**Comment by theflofly** [comment]
Some news about this issue, I am currently migrating the gradient ops from Python to C++ and I'll work on the Optimizers when enough gradients have been migrated.

**Comment by GedisJ** [comment]
@theflofly, thanks for the work you've done so far with gradients and optimizers. GradientDescentOptimizer was very useful in my project as there is no optimizer implementation in current tensorflow c++ SDK. In the future we would like to have AdamOptimizer in our software as it has proven to be more robust in our case. What are your plans on implementing that one?

**Comment by theflofly** [comment]
@GedisJ For now I am working on migrating the gradients from Python to C++ because one Operation without gradient in the graph means an untrainable network.
Because the node `​ApplyAdam` already exists I think that adding Adam is not that hard. I'll take a look and keep you updated.

**Comment by GedisJ** [comment]
@theflofly this is true about gradients, I had to add some simple gradient operations myself, before they were introduced in 1.4 release.
regarding Adam the main thing I guess is adding slots implementation to base Optimizer from what I've seen in Python code, because in C++ this seems to be missing too.

**Comment by epignatelli** [comment]
Reviving this old thread in connection with #11377.
Is there any plan / road map publicly available to keep track and contribute to this?

**Comment by theflofly** [comment]
I would also be interested.

The work I began to do was in the end discarded and I never heard back from someone at Google about working on a design...

**Comment by asimshankar** [comment]
Thanks for reviving this thread.

The sad news is that someone from Google hasn't worked on a design.
The good news is that we've gotten better in terms of establishing a process so that design changes can be made without constantly waiting on someone at Google to drive it!

If there is interest in pushing this, then it would be great for the person interested to drive the [TensorFlow RFC process](https://github.com/tensorflow/community/blob/master/governance/TF-RFCs.md), which involves writing a design proposal as a first step. This will allow the community and maintainers to see the full picture and make sure it's headed in the right direction, instead of having the design evolve through a series of one-off PRs.

I personally feel the ideal outcome would be that we'd have a path so that individual optimizer classes implemented in Python can be _migrated_ to C++ so that the Python class becomes a trivial wrapper over the functionality exposed by the C API instead of having to duplicate logic in two languages. The work earlier this year to get the Python graph building libraries to be wrappers over the C API for graph construction should make this much easier than it would have been last year.

So, long story short: No plan/roadmap available yet, but contributions are welcome to make them! :)

FYI @ewilderj

**Comment by theflofly** [comment]
Thanks for your answer @asimshankar, in the case of a RFC process do you accept being the RFC sponsor?

**Comment by ewilderj** [comment]
cc @karmel

**Comment by asimshankar** [comment]
@theflofly : I would have loved to have been, but I've since moved to other projects (and am no longer involved with TensorFlow on a daily basis) and am thus not a good person for this at the moment.

@ewilderj / @karmel should be able to find an appropriate sponsor.

**Comment by karmel** [comment]
Thanks @theflofly for keeping on us about this. If you put together an RFC for review, I can help as the sponsor to find appropriate reviewers, etc. (CC @alextp , @skye who might be good reviewers here)

**Comment by theflofly** [comment]
Thanks @asimshankar, @karmel.

I will work on something then, but bear with me I am quite busy at the moment.

**Comment by cj741** [comment]
Hi @theflofly , How is work doing? We also need the GradientDescentOptimizer in our project, but I didn't find it in the TF 2.0, hope your work merge into the TF.
